lib/sycamore/sycamore/utils/pdf_utils.py

```diff
@@ -22,6 +28,108 @@ def show_pages(docset: DocSet, limit: int = 2):
+        if isinstance(selection, list):
+            subset_start, subset_end = selection
+            if subset_end < subset_start:
+                raise ValueError("For selected_pages like [a, b] it must be that a <= b.")
+            for page_num in range(subset_start, subset_end + 1):
+                if page_num in present_pages:
+                    raise ValueError("selected_pages may not include overlapping pages.")
+                if page_num <= 0 or page_num > page_count:
+                    raise ValueError(
+                        f"Invalid page number ({page_num}): for this document,"
+                        f"page numbers must be at least 1 and at most {page_count}"
+                    )
+                present_pages.add(page_num)
+                page_list.append(page_num)
+
+                if page_num != new_page:
+                    remapped_pages[new_page] = page_num
+
+                new_page = new_page + 1
+
+        else:
+            raise ValueError("Page selection must either be an integer or a 2-element list [integer, integer]")
+    return (page_list, remapped_pages)
+
+
+def select_pdf_pages(input: BinaryIO, out: BinaryIO, page_list: list[int]) -> None:
+    input.seek(0)
+    with PdfReader(input) as pdf_reader, PdfWriter() as pdf_writer:
+        for page_num in page_list:
+            pdf_writer.add_page(pdf_reader.pages[page_num - 1])
+        pdf_writer.write_stream(out)  # see pypdf issue #2905
+    out.flush()
+
+
+def filter_elements_by_page(elements: list[Element], page_numbers: list[int]) -> list[Element]:
+    page_map = {num: idx + 1 for idx, num in enumerate(page_numbers)}
+    new_elements = []
+    for element in elements:
+        page_number = element.properties.get("page_number")
+        if page_number is not None and page_number in page_map:
+            # renumber pages so the elements reference the pages in the new document.
+            element.properties["page_number"] = page_map[page_number]
+            new_elements.append(element)
+    return new_elements
+
+
+def select_pages(page_selection: list[Union[int, list[int]]]) -> Callable[[Document], Document]:
+    """
+    Returns a function that selects pages from a PDF document based on a list of page selections.
+    Each selection can be a single page number or a range of page numbers. Page numbers are 1-indexed.
+
+    Examples:
+       [1,2,3] pages 1, 2, and 3
+       [[1,3], 5] pages 1, 2, 3, and 5
+       [[1,3], [5,7] pages 1, 2, 3, and 5, 6, 7
+       [2, 1, [4, 6]] pages 2, 1, 4, 5, 6, in that order
+
+    Args:
+       page_selection: A list of page numbers or page ranges to select. Page numbers are 1-indexed.
+
+    """
+
+    def select_pages_fn(doc: Document) -> Document:
+        if doc.binary_representation is None:
+            logging.warning("No binary_representation found in doc {doc.doc_id}. Skipping page selection.")
+            return doc
+
+        with PdfReader(BytesIO(doc.binary_representation)) as reader:
```

Comment:
I suspect this parses the entire PDF buffer, which we then do again in the call to `select_pdf_pages()`.  Is there a way to avoid this wasted work?  Perhaps a function that's just the inner part of `select_pdf_pages()`?

Reply:
Done. It's a bit more code and involved a nullcontextmanager, but seems to work fine.